libs/constrained-generators/src/Constrained/Univ.hs

```diff
@@ -30,10 +32,25 @@ class FunctionLike fn where
   -- | The semantics of a function is given by `sem`
   sem :: fn as b -> FunTy as b
 
-class Member (fn :: [Type] -> Type -> Type) fnU where
+-- | Evidence that `fn` exists somewhere in `fnU`. To make it possible
+-- for `fnU` to consist of a balanced tree of functions we need the `path`
+-- to help GHC disambiguate instances of this class during solving.
+class IsMember (fn :: [Type] -> Type -> Type) fnU (path :: [PathElem]) where
   -- | Inject a specific function symbol into a function
   -- universe.
```

Comment:
I take this is a performance improvement. Using a Tree rather than a linear list?

Reply:
Yes, it makes a big difference. Somewhere in the 10-20% range at the time when I was implementing this particular optimisation